x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedConfigReader.java

```diff
@@ -105,33 +104,33 @@ public void expandDatafeedIds(String expression, boolean allowNoDatafeeds, Clust
     public void expandDatafeedConfigs(String expression, boolean allowNoDatafeeds, ClusterState clusterState,
                                       ActionListener<List<DatafeedConfig>> listener) {
 
-        Map<String, DatafeedConfig> clusterStateConfigs = expandClusterStateDatafeeds(expression, clusterState);
-
         ExpandedIdsMatcher requiredMatches = new ExpandedIdsMatcher(expression, allowNoDatafeeds);
-        requiredMatches.filterMatchedIds(clusterStateConfigs.keySet());
 
         datafeedConfigProvider.expandDatafeedConfigsWithoutMissingCheck(expression, ActionListener.wrap(
                 datafeedBuilders -> {
-                    // Check for duplicate Ids
-                    datafeedBuilders.forEach(datafeedBuilder -> {
-                        if (clusterStateConfigs.containsKey(datafeedBuilder.getId())) {
-                            listener.onFailure(new IllegalStateException("Datafeed [" + datafeedBuilder.getId() + "] configuration " +
-                                    "exists in both clusterstate and index"));
-                            return;
-                        }
-                    });
-
                     List<DatafeedConfig> datafeedConfigs = new ArrayList<>();
                     for (DatafeedConfig.Builder builder : datafeedBuilders) {
                         datafeedConfigs.add(builder.build());
                     }
 
+                    Map<String, DatafeedConfig> clusterStateConfigs = expandClusterStateDatafeeds(expression, clusterState);
+
+                    // Duplicate configs existing in both the clusterstate and index documents are ok
+                    // this may occur during migration of configs.
+                    // Prefer the index configs and filter duplicates.
+                    for (String clusterStateDatafeedId : clusterStateConfigs.keySet()) {
+                        boolean isDuplicate = datafeedConfigs.stream()
+                                .anyMatch(datafeed -> datafeed.getId().equals(clusterStateDatafeedId));
+                        if (isDuplicate == false) {
+                            datafeedConfigs.add(clusterStateConfigs.get(clusterStateDatafeedId));
+                        }
+                    }
```

Comment:
One thing is sort of nagging me with this code:

It seems to me that if we are checking Collection Membership all the time, that it should be a `Set`, though these collections may be so small, that this sort of optimization is not necessary and may result in worse performance. So, this is a suggestion you are free to ignore if these are adequate small collections. 
```
Set<String> indexConfigIds = datafeedConfigs.stream().map(DatafeedConfig::getId).collect(Collectors.toSet());
...
if(indexConfigIds.contains(clusterStateDatafeedId) == false)
```

Reply:
I agonised over this for ages with the same nagging feeling. The code looks too complicated but  I was erring on the side that the collection would be small making any optimisation premature. I'll make your change as it's more readable

Reply:
:D, glad to see we agonized over the same thing